core/src/main/scala/cats/data/SetOps.scala

```diff
@@ -0,0 +1,125 @@
+package cats
+package data
+
+import scala.collection.immutable.SortedSet
+
+/** A set of ops methods which should be implemented for all set like
+  * structures, both the normal and non-empty variants.
+  *
+  * The purpose of this trait is merely to help ensure consistent
+  * implementation between the different concrete types.
+  *
+  * @tparam F This is the input type we are operating on. It is the result
+  *           type for operations which can never yield an empty set.
+  * @tparam G This is the the result type for operations which ''may'' yield
+  *           an empty set.
+  *
+  * The use of the types `F` and `G` allow us to share this trait between set
+  * types which may be have an empty state, and those which many not,
+  * e.g. `OrderedSet` and `NonEmptySet`.
+  */
+private[data] trait SetOps[F[_], G[_], A] {
+
+  // Abstract
+
+  def set: F[A]
+
+  def add(a: A): F[A]
+
+  def remove(a: A): G[A]
+
+  def apply(a: A): Boolean
+
+  def diff(as: F[A]): G[A]
+
+  def union(as: F[A]): F[A]
+
+  def length: Int
+
+  def foldLeft[B](b: B)(f: (B, A) => B): B
+
+  def foldRight[B](lb: Eval[B])(f: (A, Eval[B]) => Eval[B]): Eval[B]
+
+  def exists(f: A => Boolean): Boolean
+
+  def forall(f: A => Boolean): Boolean
+
+  def show(implicit A: Show[A]): String
+
+  def ===(that: F[A])(implicit A: Eq[A]): Boolean
```

Comment:
this signature is problematic for sorted sets. What if you pass an `Eq` that is in conflict with the `Order`?

I think if we keep this, `def ===(that: F[A]): Boolean` should be the signature (since all sets have either internal Order or Hash that gives them Eq).

Reply:
Good point. I'll update this if we proceed with this PR.